### ISSUE TITLE
Support Rust async programming model in SDK

### ISSUE THREAD

**Issue by tiziano88** [project-oak/oak#186]
This should allow writing idiomatic async / await code.

See
- https://doc.rust-lang.org/std/future/trait.Future.html
- https://docs.rs/tokio/0.1.22/tokio/reactor/struct.Reactor.html
- https://www.viget.com/articles/understanding-futures-in-rust-part-1/
- https://www.viget.com/articles/understanding-futures-is-rust-part-2/
- https://boats.gitlab.io/blog/post/wakers-i/
- https://boats.gitlab.io/blog/post/wakers-ii/

**Comment by tiziano88** [comment]
cc @wildarch

**Comment by tiziano88** [comment]
This may help with #592.

**Comment by blaxill** [comment]
Is this for Wasm Nodes or of the Runtime?

There is benefit for both but the implementation is less clear for the Runtime as a lot is lost from the async ecosystem without `std`. Perhaps EAL can provide the primitives necessary for the runtime to be more async-based. 

This is not a fully formed idea but a more async friendly channel model in the runtime may make it easier to add remote channels for distributed runtimes.

**Comment by tiziano88** [comment]
@daviddrysdale IIRC you mentioned you started thinking about this already, so assigning to you to flesh out what is necessary (executor, reworking examples, etc.).

Do you think we should have a sync and async version of the SDK eventually (at least for blocking stuff like reading from channels), or should everything be async?

**Comment by tiziano88** [comment]
Just came across a couple of potentially useful links:

- https://stjepang.github.io/2020/01/25/build-your-own-block-on.html
- https://github.com/spacejam/extreme

**Comment by wildarch** [comment]
Ferrous systems have done a lot of work to get async working on `no_std`, their blog posts might also make for interesting background material:
- https://ferrous-systems.com/blog/embedded-async-await/
- https://ferrous-systems.com/blog/stable-async-on-embedded/
- https://ferrous-systems.com/blog/embedded-concurrency-patterns/

Note that they focus on embedded applications, not WASM.

**Comment by tiziano88** [comment]
Nice, thanks for the links!

I hope that we should be able to reuse most of the future combinators from the `futures` crate, we mainly need to implement a `block_on` function that has `wait_on_channel` at its bottom.

**Comment by tiziano88** [comment]
@wildarch assigning to you for now, please do scope the issue a bit more precisely first, and ideally also create more specific sub-tasks for current and future work (e.g. executor, futures, combinators, parity with tokio, rewriting some of the examples, deciding whether to keep both sync and async versions, etc.)

**Comment by wildarch** [comment]
Sure, let's use this as a tracking issue and split the work into smaller parts.

One thing I noticed from the docs, `wait_on_channels` is only for read handles. I take it that writes are always non-blocking then?

**Comment by wildarch** [comment]
I did some digging to see how we could implement a futures executor for Oak using the links above, it seems it's quite tricky to implement correctly, we would need a lot of tests and/or verification to build one from scratch. 

Luckily I stumbled on the executors provided with rust `futures` crate, one that I think is of particular interest to us is [`LocalPool`](https://docs.rs/futures/0.3.5/futures/executor/struct.LocalPool.html). It has methods like [`run_until_stalled`](https://docs.rs/futures/0.3.5/futures/executor/struct.LocalPool.html#method.run_until_stalled) to poll futures until none can make more progress. With this building block building our own `block_on` impl could become as simple as:

```
// Note: This is pseudo-code
fn block_on<T>(f: Future<T>) -> T {
  let mut pool = LocalPool::new();
  let spawner = pool.spawner();

  let result_fut = spawner.spawn(f);

  loop {
    pool.run_until_stalled();
    if !has_waiting_channels() {
      break;
    }
    wait_for_channels();
  }

  pool.run_until(result_fut)
}
```

This is 100% safe Rust, and the `futures` crate is well supported by the Rust core team.

**Comment by wildarch** [comment]
I have been experimenting with an API for an asynchronous event loop (see the [hello_world](https://github.com/wildarch/oak/blob/async_sdk/examples/hello_world/module/rust/src/lib.rs) example in my branch), and I think I have an idea of what an asynchronous SDK could look like.

[This](https://github.com/wildarch/oak/blob/async_sdk/experimental/oak_async/proposal.md) proposal/RFC contains the design for the API and some of the internals. 

@tiziano88 and @daviddrysdale, what do you think? 😄

**Comment by wildarch** [comment]
First PR containing the core async primitives and executor has been merged: #1466
